geode-membership/src/main/java/org/apache/geode/distributed/internal/membership/gms/GMSMembership.java

```diff
@@ -1231,23 +1197,16 @@ public void shutdownMessageReceived(ID id, String reason) {
     if (logger.isDebugEnabled()) {
       logger.debug("Membership: recording shutdown status of {}", id);
     }
-    synchronized (this.shutdownMembers) {
-      this.shutdownMembers.put(id, id);
-      services.getHealthMonitor()
-          .memberShutdown(id, reason);
-      services.getJoinLeave().memberShutdown(id, reason);
-    }
+    this.shutdownMembers.add(id);
+    services.getJoinLeave().memberShutdown(id, reason);
```

Comment:
All threads can now concurrently invoke `memberShutdown`, is this desirable?

Reply:
My bigger concern in this version of this method is the lack of synchronization around `shutdownMembers.add(id)` since that collection is not thread-safe.

@pivotal-jbarrett I believe @kamilla1201 has a new commit coming to fix CI…

Reply:
I pushed a commit a few hours ago that makes `shutdownMembers` a `synchronizedSet`. So `shutdownMembers.add(id)` call should be thread-safe here. I'm not sure whether the concurrent invocation of memberShutdown can cause any issues.

Reply:
The other concern here is now updates to `shutdownMembers` and the call to `memberShutdown` can be concurrently interleaved.

Reply:
@pivotal-jbarrett shutdownMembers and memberShutdown are not required to be in sync.  The memberShutdown invocation informs GMSJoinLeave so it can record a LeaveRequest in case this is the membership coordinator or that it should become one.
The latestViewLock is for protecting the data used in processView() from being altered while a new view is being prepared.  The shutdownMembers field isn't used in processView() and so doesn't need to be protected by this lock.